src/arkode/arkode_io.c

```diff
@@ -1879,17 +2693,225 @@ int arkGetNumErrTestFails(void* arkode_mem, long int* netfails)
+{
+  ARKodeMem ark_mem;
+  if (arkode_mem == NULL)
+  {
+    arkProcessError(NULL, ARK_MEM_NULL, __LINE__, __func__, __FILE__,
+                    MSG_ARK_NO_MEM);
+    return (ARK_MEM_NULL);
+  }
+  ark_mem = (ARKodeMem)arkode_mem;
+
+  /* Guard against use for incompatible time stepper modules */
+  if (!ark_mem->step_supports_algebraic)
+  {
+    arkProcessError(ark_mem, ARK_STEPPER_UNSUPPORTED, __LINE__, __func__,
+                    __FILE__,
+                    "time-stepping module does not support algebraic solvers");
+    return (ARK_STEPPER_UNSUPPORTED);
+  }
+
+  /* Call stepper routine to compute the state (if provided) */
+  if (ark_mem->step_computestate)
+  {
+    return (ark_mem->step_computestate(ark_mem, zcor, z));
+  }
+  else
+  {
+    arkProcessError(ark_mem, ARK_STEPPER_UNSUPPORTED, __LINE__, __func__,
+                    __FILE__,
+                    "time-stepping module does not support this function");
+    return (ARK_STEPPER_UNSUPPORTED);
+  }
+}
+
+/*---------------------------------------------------------------
+  ARKodeGetNonlinearSystemData:
+
+  This routine provides access to the relevant data needed to
+  compute the nonlinear system function.
+  ---------------------------------------------------------------*/
+int ARKodeGetNonlinearSystemData(void* arkode_mem, sunrealtype* tcur,
+                                 N_Vector* zpred, N_Vector* z, N_Vector* Fi,
+                                 sunrealtype* gamma, N_Vector* sdata,
+                                 void** user_data)
+{
+  ARKodeMem ark_mem;
+  if (arkode_mem == NULL)
+  {
+    arkProcessError(NULL, ARK_MEM_NULL, __LINE__, __func__, __FILE__,
+                    MSG_ARK_NO_MEM);
+    return (ARK_MEM_NULL);
+  }
+  ark_mem = (ARKodeMem)arkode_mem;
+
+  /* Guard against use for incompatible time stepper modules */
+  if (!ark_mem->step_supports_algebraic)
+  {
+    arkProcessError(ark_mem, ARK_STEPPER_UNSUPPORTED, __LINE__, __func__,
+                    __FILE__,
+                    "time-stepping module does not support algebraic solvers");
+    return (ARK_STEPPER_UNSUPPORTED);
+  }
+
+  /* Call stepper routine to compute the state (if provided) */
+  if (ark_mem->step_getnonlinearsystemdata)
+  {
+    return (ark_mem->step_getnonlinearsystemdata(ark_mem, tcur, zpred, z, Fi,
+                                                 gamma, sdata, user_data));
+  }
+  else
+  {
+    arkProcessError(ark_mem, ARK_STEPPER_UNSUPPORTED, __LINE__, __func__,
+                    __FILE__,
+                    "time-stepping module does not support this function");
+    return (ARK_STEPPER_UNSUPPORTED);
+  }
+}
+
+/*---------------------------------------------------------------
+  ARKodeGetNumNonlinSolvIters:
+
+  Returns the current number of nonlinear solver iterations
+  ---------------------------------------------------------------*/
+int ARKodeGetNumNonlinSolvIters(void* arkode_mem, long int* nniters)
```

Comment:
I'm currently returning an error because this is a nonsense question for explicit steppers, however I do think that there is a stronger case for returning `0` here than in the adaptivity cases, so if others feel strongly about it I could be swayed.